File: python/mg.py
""" mg.py (not annotated for mypy, since mypy does not know frozendict) """
import frozendict, fmultiset
from listfs import * # for partition, unzip

def mrg(lst: list):
  """ merge """
  return fmultiset.fromList(lst)

def ck(labels: list) -> list:
  """ remove first features from already matched labels[0], labels[1] and T for rest """
  (nns,nps),((),pps) = labels[0],labels[1]
  return ((nns[1:],nps), ((),pps[1:])) + tuple([((),()) for label in labels[2:]])

def t(lsos:list) -> list:
  """ remove LSOs with no features in their label """
  return [x for x in lsos if x[1][1]]

def soSize(so) -> int:
  """ calculate the size of a syntactic object """
  if isinstance(so,tuple) and len(so) == 2 and \
     isinstance(so[0],tuple) and ( len(so[0]) == 0 or isinstance(so[0][0],str) ): # lex
    return 1
  elif isinstance(so,frozendict.frozendict): # fmultiset
    return 1 + sum(map(soSize, fmultiset.toList(so)))
  else:  # phtree
    return 1 + sum(map(soSize, so))

def maxx(lsos: list) -> tuple:
  """ given LSOs (of a WS), return LSO with largest SO (the head) """
  sofar = 0
  maxLSO = ((()),((),()))
  for lso in lsos:
    lsoSize = soSize(lso[0])
    if lsoSize > sofar:
      sofar = lsoSize
      maxLSO = lso
  return maxLSO

def smc(lsos: list):
  """ if LSOs have no pos feature in common, return them """
  firstPosFeatures = [lso[1][1][0] for lso in lsos if lso[1][0] == ()]
  if len(firstPosFeatures) < len(set(firstPosFeatures)):
    return fmultiset.fromList([])
  else:
    return fmultiset.fromList(lsos)

def match(wss:list) -> tuple:
  """ given list of workspaces, return ([head,complement],otherLSOs) """
  (ws0,wss0) = (wss[0],wss[1:])
  (heads, others) = partition(lambda lso: lso[1][0] != (), fmultiset.toList(ws0))
  if not(len(heads)==1): raise RuntimeError("match: 0 or >1 neg lsos in ws0")
  h = heads[0]
  f = h[1][0][0]
  if f[-1] == '+':
    f, plus = f[:-1], True
  else:
    plus = False
  (ics,iothers) = partition(lambda lso: lso[1][1][0] == f, others)
  if ics and wss0 == []:    # im
    return ([h,ics[0]], iothers)
  elif ics==[] and len(wss0)>0:   # em
    ws1 = wss0[0]
    lsos = fmultiset.toList(ws1)
    c = maxx(lsos)
    if c[1][1][0] != f:
      raise RuntimeError("match: %r != %r" % (f,c[1][1][0]))
    else:
      others1 = [x for x in lsos if x != c]
      if plus and others == others1:
        return ([h,c]+atb(f, others, wss0[1:]), others)
      elif wss0[1:] == []:
        return ([h,c], others + others1)
      else:
        raise RuntimeError("match: too many wss")

def atb(f, movers, wss):
  """ collect comps with first feature f and others==movers """
  if wss == []:
    return []
  else:
    ws1 = wss[0]
    lsos = fmultiset.toList(ws1)
    c = maxx(lsos)
    if c[1][1][0] != f:
      raise RuntimeError("match: complement clash")
    else:
      others = [x for x in lsos if x != c]
      if others != movers:
        raise RuntimeError("match: mover clash")
      else:
        return [c] + atb(f, movers, wss[1:])

def d(wss:list):
  """ the derivational step: given list of workspaces, return derived workspace """
  (matched, movers) = match(wss)
  (sos,labels) = unzip(matched)
  return smc( t( list(zip( [mrg(sos)]+sos[1:], ck(labels))) + movers))

File: haskell/Mg.hs
module Mg where     -- Multiset needed. E.g., start ghci with: stack ghci --package multiset
import Data.MultiSet (MultiSet)
import qualified Data.MultiSet as MultiSet
import Data.List
import qualified Data.List as List

data F = C | D | N | V | A | P | Wh | Pred | Predx | T | K | Vx | Scr |
         Modal | Have | Be | Been | Ving | Ven deriving (Show, Eq, Ord)
data Ft = One F | Plus F deriving (Show, Eq, Ord) -- build features from base F
type Label = ([Ft], [Ft])
type Lex = ([String], Label)
data PhTree = Pl Lex | Ps [PhTree] deriving (Show, Eq, Ord)
data SO = L Lex | S (MultiSet (SO)) | O PhTree deriving (Show, Eq, Ord)
type LSO = (SO, Label)
type WS = MultiSet (LSO)

mrg :: [SO] -> SO
mrg sos = S (MultiSet.fromList sos)

ck :: [Label] -> [Label] -- already matched features can be `forgotten'
ck ((_:nns,nps):([],_:pps):more) = [(nns,nps), ([],pps) ] ++ (map (\label -> ([],[])) more)

t :: [LSO] -> [LSO]      -- constituents with no remembered features can be `forgotten'
t = filter (\lso -> snd lso /= ([],[]))

soSize :: SO -> Int
soSize (S so) = foldr (\x y -> (soSize x) + y) 1 (MultiSet.toList so)
soSize (O (Ps ts)) = foldr (\x y -> (soSize (O x)) + y) 1 ts
soSize _ = 1

maxx :: [LSO] -> LSO
maxx lsos = foldr1 (\x y ->if ((soSize.fst) x) >= ((soSize.fst) y) then x else y) lsos

match:: [WS] -> ([LSO],[LSO])
match (ws0:wss) = case List.partition ((/= []).fst.snd) (MultiSet.toList ws0) of
    ([h],others) -> let (f,plus) = (fplus.head.fst.snd) h in
      case (List.partition ((== (One f)).head.snd.snd) others,wss) of
        (([c],others'),[]) -> ([h,c],others')                                        -- IM
        (([],_),ws1:wss) -> let lsos = (MultiSet.toList ws1) in let c = maxx lsos in -- EM
          if ((/= One f).head.snd.snd) c
          then error "match complement feature clash"
          else let others' = List.filter (/= c) lsos in
            if plus && others' == others
            then (h:c:(atb (One f) others wss), others)
            else if wss == [] then ([h,c], others ++ others') else (error ("match: too many wss"))
  where
    fplus :: Ft -> (F, Bool)             -- parse the one/plus features
    fplus ft = case ft of (One f) -> (f, False); (Plus f) -> (f, True)

    atb :: Ft -> [LSO] -> [WS] -> [LSO]  -- collect comps with first feature f, and others
    atb _ _ [] = []
    atb ft movers (ws:wss) = case List.partition ((== ft).head.snd.snd) (MultiSet.toList ws) of
        ([c'],others) -> if others == movers then c':(atb ft movers wss) else (error "match: ATB error")

smc :: [LSO] -> WS
smc lsos = if smc' [] lsos then MultiSet.fromList lsos else error "smc violation"
  where
    smc' _ [] = True
    smc' sofar ((s,([],p:ps)):lsos) = if elem p sofar then False else smc' (p:sofar) lsos
    smc' sofar (_:lsos) = smc' sofar lsos

-- the derivational step: unbounded merge and label
d :: [WS] -> WS
d wss = let (matched,movers) = match wss in
        let (sos,labels) = unzip matched in
           smc (t (zip (mrg sos:tail sos) (ck labels) ++ movers))
